File: C++/N-Queens.cc
// The n-queens puzzle is the problem of placing n queens on an n×n chessboard such that no two queens
// attack each other.

// Given an integer n, return all distinct solutions to the n-queens puzzle.

// Each solution contains a distinct board configuration of the n-queens' placement, where 'Q' and '.'
// both indicate a queen and an empty space respectively.

// For example,
// There exist two distinct solutions to the 4-queens puzzle:

// [
//  [".Q..",  // Solution 1
//   "...Q",
//   "Q...",
//   "..Q."],

//  ["..Q.",  // Solution 2
//   "Q...",
//   "...Q",
//   ".Q.."]
// ]

class Solution {
    bool isValid(int i, int j, vector<pair<int, int>>& queens) {
        for (auto& p : queens) {
            if (j == p.second || abs(i - p.first) == abs(j - p.second)) {
                return false;
            }
        }
        return true;
    }

    void solve(vector<string>& cur, int row, int count, vector<pair<int, int>>& queens, vector<vector<string>>& output) {
        if (count == 0) {
            output.push_back(cur);
            return;
        }

        for (int j = 0; j < cur[0].size(); j++) {
            if (isValid(row, j, queens)) {
                cur[row][j] = 'Q';
                queens.emplace_back(row, j);
                solve(cur, row + 1, count - 1, queens, output);
                cur[row][j] = '.';
                queens.pop_back();
            }
        }

    }
public:
    vector<vector<string>> solveNQueens(int n) {
        vector<string> cur(n, string(n, '.'));
        vector<pair<int, int>> queens;
        vector<vector<string>> output;
        solve(cur, 0, n, queens, output);
        return output;
    }
};

File: Java/N-queens.java
// The n-queens puzzle is the problem of placing n queens on an n×n chessboard such that no two queens 
// attack each other.

// Given an integer n, return all distinct solutions to the n-queens puzzle.

// Each solution contains a distinct board configuration of the n-queens' placement, where 'Q' and '.' both 
// indicate a queen and an empty space respectively.

// For example,
// There exist two distinct solutions to the 4-queens puzzle:

// [
//  [".Q..",  // Solution 1
//   "...Q",
//   "Q...",
//   "..Q."],

//  ["..Q.",  // Solution 2
//   "Q...",
//   "...Q",
//   ".Q.."]
// ]

public class N-queens {
    List<String> buildBoard(char[][] board) {
        List<String> ls = new ArrayList<>();
        for (char[] row : board) {
            ls.add(new String(row));
        }
        return ls;
    }
    
    boolean isValid(char[][] board, int row, int col) {
        for (int i = 0; i < board.length; i++) {
            if (board[i][col] == 'Q') {
                return false;
            }
        }
        
        for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; i--, j--) {
            if (board[i][j] == 'Q') {
                return false;
            }
        }
        
        for (int i = row - 1, j = col + 1; i >= 0 && j < board.length; i--, j++) {
            if (board[i][j] == 'Q') {
                return false;
            }
        }
        return true;
    }
    
    void solveNQueens(char[][] board, int row, List<List<String>> out) {
            if (row == board.length) {
                out.add(buildBoard(board));
                return;
            }
            for (int j = 0; j < board.length; j++) {
                if (isValid(board, row, j)) {
                    board[row][j] = 'Q';
                    solveNQueens(board, row + 1, out);
                    board[row][j] = '.';
                }
            }
        }
    
    public List<List<String>> solveNQueens(int n) {
        List<List<String>> out = new ArrayList<>();
        char[][] board = new char[n][n];
        for (int i = 0; i < board.length; i++) {
            Arrays.fill(board[i], '.');
        }
        solveNQueens(board, 0, out);
        return out;
    }
}
